databuilder/extractor/bigquery_usage_extractor.py

```diff
@@ -0,0 +1,103 @@
+from collections import namedtuple
+from datetime import date, timedelta
+import logging
+
+from googleapiclient.discovery import build
+from oauth2client.client import GoogleCredentials
+
+from databuilder.extractor.base_extractor import Extractor
+
+TableColumnUsageTuple = namedtuple('TableColumnUsageTuple', ['database', 'cluster', 'schema',
+                                                             'table', 'column', 'email'])
+
+LOGGER = logging.getLogger(__name__)
+
+
+class BigQueryUsageExtractor(Extractor):
+    """
+    An aggregate extractor for bigquery table usage. This class takes the data from
+    the stackdriver logging API by filtering on timestamp, bigquery_resource and looking
+    for referencedTables in the response.
+    """
+    TIMESTAMP_KEY = 'timestamp'
+    PROJECT_ID_KEY = 'projectid'
+
+    def init(self, conf):
+        credentials = GoogleCredentials.get_application_default()
+        self.logging_service = build('logging', 'v2', credentials=credentials)
+
+        # type: (ConfigTree) -> None
+        self.timestamp = conf.get_string(
+            BigQueryUsageExtractor.TIMESTAMP_KEY,
+            (date.today() - timedelta(days=1)).strftime('%Y-%m-%dT00:00:00Z'))
+        self.projectid = conf.get_string(BigQueryUsageExtractor.PROJECT_ID_KEY)
+        self.table_usage_counts = {}
+        self._count_usage()
+        self.iter = iter(self.table_usage_counts)
+
+    def _count_usage(self):
+        for entry in self._retrieve_records():
+            if not isinstance(entry, dict):
+                raise Exception("BigQueryUsageExtractor expects record of type dict!")
+
+            email = entry['protoPayload']['authenticationInfo']['principalEmail']
+            jobCompletedEvent = entry['protoPayload']['serviceData']['jobCompletedEvent']
+            refTables = jobCompletedEvent['job']['jobStatistics'].get('referencedTables', None)
+
+            if not refTables:
+                # Query results can be cached and if the source tables remain untouched,
+                # bigquery will return it from a 24 hour cache result instead.
+                continue
+
+            for refTable in refTables:
+                key = TableColumnUsageTuple(database='bigquery',
+                                            cluster=refTable['projectId'],
+                                            schema=refTable['datasetId'],
+                                            table=refTable['tableId'],
+                                            column='*',
```

Comment:
so now it only considers table usage not column usage?

Reply:
Yes, table level only for now. To make this column level would require parsing the query. That is also available, but requires a good amount of extra work. There is a WIP grammar for it though: https://github.com/Andrew0Hill/antlr-bq-grammar

Reply:
> Yes, table level only for now. To make this column level would require parsing the query. That is also available, but requires a good amount of extra work. There is a WIP grammar for it though: https://github.com/Andrew0Hill/antlr-bq-grammar

Once we have clear case for column usage in Amundsen, and BQ Antlr grammar is stable, I guess we can consolidate this into `TblColUsgAggExtractor`

Reply:
thanks @gtoonstra . Make sense. We could add it later.